src/compiler/AST/traverse.js

```diff
@@ -1,22 +1,9 @@
 // @flow
 
-function cloneNode(o: Node): Node {
-  let v, key;
-  const output = Array.isArray(o) ? [] : {};
-  for (key in o) {
-    // $FlowIgnore
-    v = o[key];
-    // $FlowIgnore
-    output[key] = (typeof v === 'object') ? cloneNode(v) : v;
-  }
-  // $FlowIgnore
-  return output;
-}
-
 function createPath(node: Node): NodePath<Node> {
 
   return {
-    node: cloneNode(node),
+    node,
```

Comment:
Well, I tried to use an immutable tree but didn't succeed. I hope we won't end up in the same state as Babel with unexpected mutations. I think it will be fine since we won't have much transformations like this here.